### DISCUSSION TITLE
[SPRING BOOT] @SpringBootApplication 알아보기

### DISCUSSION THREAD

**Discussion by ahngilwoong** [cs-collections/cs-collections-discussion#15] [Spring]
### SpringBootApplication은 어떻게 이루어져 있을까?

**_@SpringBootApplication의 대표적인 3가지 어노테이션_**
1. @SpringBootConfiguration
2. @ ComponentScan
3. @EnableAutoConfiguration
<img width="444" alt="image" src="https://user-images.githubusercontent.com/65801290/208836999-93770eb3-f7d7-4128-b6d5-a0376f466398.png">

### 1. @SpringBootConfiguration
SpringBootConfiguration은 SpringBoot만의 Configuration이며 해당 클래스가 @Bean 메서드를 정의되어 있음을 
Spring 컨테이너에 알려주는 역할을 한다.
@Configuration과의 차이는 구성을 자동으로 찾을 수 있느냐의 차이이다.
클래스에 @Configuration이 있으면 스프링은 Java Configure 설정 클래스로 간주하여 Bean으로 등록한다.
@SpringBootConfiguration 어노테이션은 애플리케이션에 1개만 존재해야하고 
이 어노테이션을 기준으로 설정들을 다 불러오게 된다.

### 2. @ ComponentScan (개발자가 만든 빈들을 등록하는 역할)
ComponentScan의 base-package로 지정한 패키지 하위의 모든 @Component가 붙은 것들을 스캔하여 빈으로 등록한다.
@Controller, @Service, @Repository, @RestContoroller, @Component, @Configuration가 있는 것들이 대상이다.
위 어노테이션들을  내부 구현으로 까보면 @Component가 정의되어있다. 
base-package가 정의되지 않으면 해당 어노테이션이 붙은 classpath 하위의 @Component 어노테이션을 스캔해서, Bean으로 등록한다.

### 3. @EnableAutoConfiguration (Spring 또는 라이브러리에서 미리 정의된 클래스들을 빈으로 등록하는 역할)
미리 정의된 라이브러리들을  Bean으로 등록해 주는 어노테이션이다.
미리 정의된 라이브러리들을 모두 등록하지는 않고  Conditional이 만족될 경우에 Bean으로 등록된다.
즉 Spring 응용 프로그램 컨텍스트의 자동 구성을 활성화하고 미리 정의된 라이브러리를 Bean으로 등록시켜주는 어노테이션이다.
Tomcat, DispatcherServlet 자동설정 등 많은 수의 자동설정을 기본으로 제공한다.
Dependencies > spring-boot-autoconfigure > META-INF > spring.factories (미리 정의된 라이브러리들이 있는 곳)

**Comment by Irisation23** [comment]
정리 해주신 글 잘 읽었습니다.
`@ComponentScan` 과  `@EnableAutoConfiguration`의 어노테이션 정리도 
**그림**이나 **스크린샷** 첨부 해 주시면 정리하는데 도움이 될 것 같습니다.
